src/components/ConnectWallet/ConnectWalletDialog.tsx

```diff
@@ -0,0 +1,229 @@
+import { Header } from 'components/Dialog'
+import Row from 'components/Row'
+import EventEmitter from 'events'
+import { connections, defaultChainIdAtom, Web3Connection } from 'hooks/connectWeb3/useWeb3React'
+import { atom, useAtom } from 'jotai'
+import { useAtomValue } from 'jotai/utils'
+import QRCode from 'qrcode'
+import { useCallback, useEffect, useState } from 'react'
+import styled from 'styled-components/macro'
+import { lightTheme, ThemedText } from 'theme'
+
+const Body = styled(Column)`
+  height: calc(100% - 2.5em);
+`
+
+const SecondaryOptionsRow = styled(Row)`
+  align-self: end;
+  grid-template-columns: repeat(2, calc(50% - 0.75em / 2));
+  height: fit-content;
+`
+
+const ButtonContents = styled(Column)`
+  gap: 0.75em;
+  justify-items: center;
+`
+
+const StyledMainButton = styled(Button)`
+  background-color: ${({ theme }) => theme.container};
+  border-radius: ${({ theme }) => theme.borderRadius * 0.75}em;
+  height: 183px;
+  padding: 22px;
+`
+
+const StyledMainButtonRow = styled(Row)`
+  grid-template-columns: repeat(2, calc(50% - 1em / 2));
+  justify-items: center;
+`
+
+const StyledSmallButton = styled(Button)`
+  background-color: ${({ theme }) => theme.container};
+  border-radius: ${({ theme }) => theme.borderRadius * 0.75}em;
+  height: 90px;
+  padding: 16px;
+`
+
+const StyledNoWalletText = styled(ThemedText.Subhead1)`
+  line-height: 20px;
+  white-space: pre-wrap;
+`
+
+const QRCodeWrapper = styled.div`
+  height: 110px;
+  width: 110px;
+  path {
+    /* Maximize contrast: transparent in light theme, otherwise hard-coded to light theme. */
+    fill: ${({ theme }) => (theme.container === lightTheme.container ? '#00000000' : lightTheme.container)};
+  }
+`
+
+interface ButtonProps {
+  walletName?: string
+  logoSrc?: string
+  connection?: Web3Connection
+  onClick: () => void
+  onError?: (e: Error | undefined) => void
+}
+
+const wcQRUriAtom = atom<string | undefined>(undefined)
+
+function toQrCodeSvg(qrUri: string): Promise<string> {
+  return QRCode.toString(qrUri, {
+    // Leave a margin to increase contrast in dark mode.
+    margin: 1,
+    // Use 55*2=110 for the width to prevent distortion. The generated viewbox is "0 0 55 55".
+    width: 110,
+    type: 'svg',
+  })
+}
+
+function WalletConnectButton({ walletName, logoSrc, connection: wcTileConnection, onClick, onError }: ButtonProps) {
+  const [walletConnect] = wcTileConnection as [WalletConnect, Web3ReactHooks]
+  const defaultChainId = useAtomValue(defaultChainIdAtom)
+
+  const [qrUri, setQrUri] = useAtom(wcQRUriAtom)
+  const [qrCodeSvg, setQrCodeSvg] = useState<string>('')
+
+  useEffect(() => {
+    let stale = false
+    if (qrUri) {
+      toQrCodeSvg(qrUri).then((qrCodeSvg) => {
+        if (stale) return
+        setQrCodeSvg(qrCodeSvg)
+      })
+    } else {
+      walletConnect.activate(defaultChainId).catch((e) => {
+        if (stale) return
+        onError?.(e)
+      })
+    }
+    return () => {
```

Comment:
what does returning a function in a use effect do?

Reply:
https://reactjs.org/docs/hooks-effect.html#effects-with-cleanup